lib/offline/storage.js

```diff
@@ -770,6 +788,61 @@ shaka.offline.Storage.prototype.createOfflineManifest_ = function(
   };
 };
 
+/**
+ * Extract DRM initData from first segment
+ *
+ * @param {shaka.extern.Manifest} manifest
+ * @param {shaka.net.NetworkingEngine} networkingEngine
+ * @return {Promise<shaka.extern.Manifest>}
+ * @private
+ */
+shaka.offline.Storage.prototype.extractInitDataFromFirstSegment_ = async function(
```

Comment:
This would probably be best done in `shaka.offline.DownloadManager` since we are downloading the segments anyway.  Plus you should check every segment since there can be `pssh` boxes in following segments for key rotation.

Reply:
> This would probably be best done in shaka.offline.DownloadManager since we are downloading the segments anyway. 
I was not 100% sure if the initial segment was actually downloaded for offline usage, understanding one could potentially hook into the existing download stream we faced the issue that the DRMEngine was initialised before the download manager is started.
As we did not directly had a good reference how to inject a PSSH key after the DRM engine was already initialised (the engine failed on not having a pssh value), we found it easier to inject it before the DRM engine was loaded so we condensed the logic into storage.js and injecting the pssh data into the manifest so it would be seamless work with the existing shaka code

Per above we could break up the method and move part of the logic to the downloadManager, however it might still be easier to download the first segments separately first to prevent larger adjustments to classes like the DrmEngine.

> Plus you should check every segment since there can be pssh boxes in following segments for key rotation.
We are currently not working with content that is using key rotation, would that even work for online content? But are happy to contribute. Can you point us in the right direction how this could conceptually be applied in this setup?

Reply:
It's fine that DrmEngine is initialized first, you just need to remove the [error that is thrown](https://github.com/google/shaka-player/blob/master/lib/offline/storage.js#L800) because of missing init data.  Then you can use `drmEngine.newInitData` from DownloadManager to signal DrmEngine to create a new session.  Lastly, update the offline manifest object before storing it with the new session IDs.